fix(compiler-cli): disable tree shaking during HMR

crisbeto opened the pull request:
When HMR is enabled, we need to capture the dependencies used in a template and forward them to the HMR replacement function. One half of this process is static, meaning that we can't change it after the initial compilation. Tree shaking becomes a problem in such a case, because the user can change the template in a way that changes the set of dependencies which will start mismatching with the static part of the HMR code.

These changes disable the tree shaking when HMR is enabled to ensure that the dependencies stay stable.

Fixes #59581.

msmallest commented on packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts:
```diff
@@ -1222,14 +1222,25 @@ export class ComponentDecoratorHandler
-          continue;
+
+      if (this.enableHmr) {
+        // In HMR we need to preserve all the dependencies, because they have to remain consitent
```
nit: spelling of comment

```suggestion
        // In HMR we need to preserve all the dependencies, because they have to remain consistent
```

crisbeto committed:
fix(compiler-cli): disable tree shaking during HMR

When HMR is enabled, we need to capture the dependencies used in a template and forward them to the HMR replacement function. One half of this process is static, meaning that we can't change it after the initial compilation. Tree shaking becomes a problem in such a case, because the user can change the template in a way that changes the set of dependencies which will start matching with the static part of the HMR code.

These changes disable the tree shaking when HMR is enabled to ensure that the dependencies stay stable.

Fixes #59581.

pkozlowski-opensource commented:
This PR was merged into the repository by commit f9b13e4e58b5499e0108df841e9f96ca10e0a667.

The changes were merged into the following branches: main, 19.1.x
